### ISSUE TITLE
java.util.NoSuchElementException: head of empty list

### ISSUE THREAD

**Issue by beaushinkle** [etorreborre/specs2#501]
I'm getting:
```
[error]
[error] java.util.NoSuchElementException: head of empty list
[error]
[error] STACKTRACE
[error]   scala.collection.immutable.Nil$.head(List.scala:420)
[error]   scala.collection.immutable.Nil$.head(List.scala:417)
[error]   org.specs2.foldm.stream.FoldProcessM$$anon$4$$anonfun$fold$1$$anonfun$apply$1.apply(FoldProcessM.scala:34)
[error]   org.specs2.foldm.stream.FoldProcessM$$anon$4$$anonfun$fold$1$$anonfun$apply$1.apply(FoldProcessM.scala:34)
[error]   scalaz.concurrent.Task$$anonfun$flatMap$1$$anonfun$1.apply(Task.scala:35)
[error]   scalaz.concurrent.Task$$anonfun$flatMap$1$$anonfun$1.apply(Task.scala:35)
[error]   scalaz.concurrent.Task$.Try(Task.scala:448)
[error]   scalaz.concurrent.Task$$anonfun$flatMap$1.apply(Task.scala:35)
[error]   scalaz.concurrent.Task$$anonfun$flatMap$1.apply(Task.scala:33)
[error]   scalaz.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:58)
[error]   scalaz.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:58)
[error]   scalaz.concurrent.Future.step(Future.scala:118)
[error]   scalaz.concurrent.Future.unsafePerformListen(Future.scala:75)
[error]   scalaz.concurrent.Future.unsafePerformAsync(Future.scala:159)
[error]   scalaz.concurrent.Future.unsafePerformSync(Future.scala:184)
[error]   scalaz.concurrent.Task.unsafePerformSync(Task.scala:99)
[error]   scalaz.concurrent.Task.run(Task.scala:106)
[error]   org.specs2.foldm.stream.FoldProcessM$$anon$4$$anonfun$fold$1.apply(FoldProcessM.scala:34)
[error]   org.specs2.foldm.stream.FoldProcessM$$anon$4$$anonfun$fold$1.apply(FoldProcessM.scala:33)
[error]   org.specs2.foldm.FoldM$$anon$10$$anonfun$fold$3.apply(FoldM.scala:114)
[error]   org.specs2.foldm.FoldM$$anon$10$$anonfun$fold$3.apply(FoldM.scala:114)
[error]   org.specs2.foldm.FoldM$$anon$10$$anonfun$fold$3.apply(FoldM.scala:114)
[error]   org.specs2.foldm.FoldM$$anon$10$$anonfun$fold$3.apply(FoldM.scala:114)
[error]   org.specs2.foldm.FoldM$$anon$10$$anonfun$fold$3.apply(FoldM.scala:114)
[error]   org.specs2.foldm.FoldM$$anon$10$$anonfun$fold$3.apply(FoldM.scala:114)
[error]   org.specs2.foldm.stream.FoldableProcessM$$anon$1$$anonfun$org$specs2$foldm$stream$FoldableProcessM$$anon$$go$1$1.apply(FoldableProcessM.scala:19)
[error]   org.specs2.foldm.stream.FoldableProcessM$$anon$1$$anonfun$org$specs2$foldm$stream$FoldableProcessM$$anon$$go$1$1.apply(FoldableProcessM.scala:18)
[error]   scalaz.$bslash$div.fold(Either.scala:58)
[error]   org.specs2.codata.Process$$anonfun$awaitOr$1$$anonfun$apply$48$$anonfun$apply$49.apply(Process.scala:743)
[error]   org.specs2.codata.Process$$anonfun$awaitOr$1$$anonfun$apply$48$$anonfun$apply$49.apply(Process.scala:743)
[error]   org.specs2.codata.Util$.Try(Util.scala:38)
[error]   org.specs2.codata.Process$$anonfun$awaitOr$1$$anonfun$apply$48.apply(Process.scala:743)
[error]   org.specs2.codata.Process$$anonfun$awaitOr$1$$anonfun$apply$48.apply(Process.scala:743)
[error]   scalaz.Trampoline$$anonfun$delay$1.apply(Free.scala:338)
[error]   scalaz.Trampoline$$anonfun$delay$1.apply(Free.scala:338)
[error]   scalaz.Free$$anonfun$suspend$2.apply(Free.scala:27)
[error]   scalaz.Free$$anonfun$suspend$2.apply(Free.scala:27)
[error]   scalaz.std.FunctionInstances$$anon$1$$anonfun$map$1.apply(Function.scala:77)
[error]   scalaz.Free$$anonfun$run$1.apply(Free.scala:251)
[error]   scalaz.Free$$anonfun$run$1.apply(Free.scala:251)
[error]   scalaz.Free.go2$1(Free.scala:156)
[error]   scalaz.Free.go(Free.scala:159)
[error]   scalaz.Free.run(Free.scala:251)
[error]   org.specs2.codata.process1$Await1$$anonfun$unapply$1$$anonfun$apply$39.apply(process1.scala:841)
[error]   org.specs2.codata.process1$Await1$$anonfun$unapply$1$$anonfun$apply$39.apply(process1.scala:841)
[error]   org.specs2.codata.Util$.Try(Util.scala:38)
[error]   org.specs2.codata.process1$Await1$$anonfun$unapply$1.apply(process1.scala:841)
[error]   org.specs2.codata.process1$Await1$$anonfun$unapply$1.apply(process1.scala:841)
[error]   org.specs2.codata.process1$.go$6(process1.scala:197)
[error]   org.specs2.codata.process1$.feed(process1.scala:202)
[error]   org.specs2.codata.Process$$anonfun$pipe$1.apply(Process.scala:145)
[error]   org.specs2.codata.Process$$anonfun$pipe$1.apply(Process.scala:139)
[error]   org.specs2.codata.Process$$anonfun$flatMap$1.apply(Process.scala:46)
[error]   org.specs2.codata.Process$$anonfun$flatMap$1.apply(Process.scala:46)
[error]   org.specs2.codata.Util$.Try(Util.scala:38)
[error]   org.specs2.codata.Process$class.flatMap(Process.scala:46)
[error]   org.specs2.codata.Process$Emit.flatMap(Process.scala:592)
[error]   org.specs2.codata.Process$$anonfun$flatMap$4.apply(Process.scala:48)
[error]   org.specs2.codata.Process$$anonfun$flatMap$4.apply(Process.scala:48)
[error]   scalaz.Free$$anonfun$map$1.apply(Free.scala:91)
[error]   scalaz.Free$$anonfun$map$1.apply(Free.scala:91)
[error]   scalaz.Free.resume(Free.scala:109)
[error]   scalaz.Free.go2$1(Free.scala:155)
[error]   scalaz.Free.go(Free.scala:159)
[error]   scalaz.Free.run(Free.scala:251)
[error]   org.specs2.codata.Process$$anonfun$go$1$1.apply(Process.scala:84)
[error]   org.specs2.codata.Process$$anonfun$go$1$1.apply(Process.scala:84)
[error]   org.specs2.codata.Util$.Try(Util.scala:38)
[error]   org.specs2.codata.Process$class.go$1(Process.scala:84)
[error]   org.specs2.codata.Process$class.step(Process.scala:97)
[error]   org.specs2.codata.Process$Append.step(Process.scala:641)
[error]   org.specs2.codata.Process$$anonfun$pipe$1.apply(Process.scala:143)
[error]   org.specs2.codata.Process$$anonfun$pipe$1.apply(Process.scala:139)
[error]   org.specs2.codata.Process$$anonfun$flatMap$1.apply(Process.scala:46)
[error]   org.specs2.codata.Process$$anonfun$flatMap$1.apply(Process.scala:46)
[error]   org.specs2.codata.Util$.Try(Util.scala:38)
[error]   org.specs2.codata.Process$class.flatMap(Process.scala:46)
[error]   org.specs2.codata.Process$Emit.flatMap(Process.scala:592)
[error]   org.specs2.codata.Process$$anonfun$flatMap$4.apply(Process.scala:48)
[error]   org.specs2.codata.Process$$anonfun$flatMap$4.apply(Process.scala:48)
[error]   scalaz.Free$$anonfun$map$1.apply(Free.scala:91)
[error]   scalaz.Free$$anonfun$map$1.apply(Free.scala:91)
[error]   scalaz.Free.resume(Free.scala:109)
[error]   scalaz.Free.go2$1(Free.scala:155)
[error]   scalaz.Free.go(Free.scala:159)
[error]   scalaz.Free.run(Free.scala:251)
[error]   org.specs2.codata.Process$$anonfun$go$1$1.apply(Process.scala:84)
[error]   org.specs2.codata.Process$$anonfun$go$1$1.apply(Process.scala:84)
[error]   org.specs2.codata.Util$.Try(Util.scala:38)
[error]   org.specs2.codata.Process$class.go$1(Process.scala:84)
[error]   org.specs2.codata.Process$class.step(Process.scala:97)
[error]   org.specs2.codata.Process$Append.step(Process.scala:641)
[error]   org.specs2.codata.Process$class.go$3(Process.scala:482)
[error]   org.specs2.codata.Process$$anonfun$go$3$3.apply(Process.scala:490)
[error]   org.specs2.codata.Process$$anonfun$go$3$3.apply(Process.scala:490)
[error]   scalaz.concurrent.Task$$anonfun$flatMap$1$$anonfun$1.apply(Task.scala:35)
[error]   scalaz.concurrent.Task$$anonfun$flatMap$1$$anonfun$1.apply(Task.scala:35)
[error]   scalaz.concurrent.Task$.Try(Task.scala:448)
[error]   scalaz.concurrent.Task$$anonfun$flatMap$1.apply(Task.scala:35)
[error]   scalaz.concurrent.Task$$anonfun$flatMap$1.apply(Task.scala:33)
[error]   scala.Function1$$anonfun$andThen$1.apply(Function1.scala:52)
[error]   scalaz.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:58)
[error]   scalaz.concurrent.Future$$anonfun$flatMap$1.apply(Future.scala:58)
[error]   scalaz.concurrent.Future.step(Future.scala:118)
[error]   scalaz.concurrent.Future.unsafePerformListen(Future.scala:75)
[error]   scalaz.concurrent.Future.unsafePerformAsync(Future.scala:159)
[error]   scalaz.concurrent.Future.unsafePerformSync(Future.scala:184)
[error]   scalaz.concurrent.Task.unsafePerformSyncAttempt(Task.scala:110)
[error]   scalaz.concurrent.Task.attemptRun(Task.scala:114)
[error]   org.specs2.control.package$taskToAction.toAction(package.scala:131)
[error]   org.specs2.control.package$Actions$.fromTask(package.scala:162)
[error]   org.specs2.reporter.Reporter$$anonfun$report$1.apply(Reporter.scala:54)
[error]   org.specs2.reporter.Reporter$$anonfun$report$1.apply(Reporter.scala:42)
[error]   org.specs2.runner.Runner$.runSpecStructure(Runner.scala:102)
[error]   org.specs2.runner.SbtRunner$$anonfun$specificationRun$1.apply(SbtRunner.scala:85)
[error]   org.specs2.runner.SbtRunner$$anonfun$specificationRun$1.apply(SbtRunner.scala:84)
[error]   org.specs2.control.eff.Arrs.go$2(Eff.scala:243)
[error]   org.specs2.control.eff.Arrs.apply(Eff.scala:253)
[error]   org.specs2.control.eff.Interpret$$anon$1$$anonfun$onEffect$1.apply(Interpret.scala:53)
[error]   org.specs2.control.eff.Interpret$$anon$1$$anonfun$onEffect$1.apply(Interpret.scala:53)
[error]   scalaz.$bslash$div.bimap(Either.scala:91)
[error]   org.specs2.control.eff.Interpret$$anon$1.onEffect(Interpret.scala:53)
[error]   org.specs2.control.eff.Interpret$$anon$1.onEffect(Interpret.scala:45)
[error]   org.specs2.control.eff.Interpret$class.go$1(Interpret.scala:131)
[error]   org.specs2.control.eff.Interpret$class.interpretLoop(Interpret.scala:142)
[error]   org.specs2.control.eff.Interpret$.interpretLoop(Interpret.scala:230)
[error]   org.specs2.control.eff.Interpret$class.interpret(Interpret.scala:55)
[error]   org.specs2.control.eff.Interpret$.interpret(Interpret.scala:230)
[error]   org.specs2.control.eff.Interpret$class.interpret1(Interpret.scala:62)
[error]   org.specs2.control.eff.Interpret$.interpret1(Interpret.scala:230)
[error]   org.specs2.control.eff.ErrorInterpretation$class.runError(ErrorEffect.scala:77)
[error]   org.specs2.control.eff.ErrorEffect$.runError(ErrorEffect.scala:129)
[error]   org.specs2.control.eff.syntax.error$ErrorEffectOps.runError(error.scala:15)
[error]   org.specs2.control.package$.executeAction(package.scala:48)
[error]   org.specs2.runner.SbtRunner$$anon$4$$anonfun$execute$1.apply(SbtRunner.scala:57)
[error]   org.specs2.runner.SbtRunner$$anon$4$$anonfun$execute$1.apply(SbtRunner.scala:55)
[error]   scala.Option.foreach(Option.scala:257)
[error]   org.specs2.runner.SbtRunner$$anon$4.execute(SbtRunner.scala:55)
[error]   sbt.TestRunner.runTest$1(TestFramework.scala:76)
[error]   sbt.TestRunner.run(TestFramework.scala:85)
[error]   sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
[error]   sbt.TestFramework$$anon$2$$anonfun$$init$$1$$anonfun$apply$8.apply(TestFramework.scala:202)
[error]   sbt.TestFramework$.sbt$TestFramework$$withContextLoader(TestFramework.scala:185)
[error]   sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
[error]   sbt.TestFramework$$anon$2$$anonfun$$init$$1.apply(TestFramework.scala:202)
[error]   sbt.TestFunction.apply(TestFramework.scala:207)
[error]   sbt.Tests$.sbt$Tests$$processRunnable$1(Tests.scala:239)
[error]   sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
[error]   sbt.Tests$$anonfun$makeSerial$1.apply(Tests.scala:245)
[error]   sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
[error]   sbt.std.Transform$$anon$3$$anonfun$apply$2.apply(System.scala:44)
[error]   sbt.std.Transform$$anon$4.work(System.scala:63)
[error]   sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
[error]   sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:228)
[error]   sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
[error]   sbt.Execute.work(Execute.scala:237)
[error]   sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
[error]   sbt.Execute$$anonfun$submit$1.apply(Execute.scala:228)
[error]   sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
[error]   sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
[error]   java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]   java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
[error]   java.util.concurrent.FutureTask.run(FutureTask.java:266)
[error]   java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
[error]   java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
[error]   java.lang.Thread.run(Thread.java:745)
[error]
[error]
[error] This looks like a specs2 exception...
[error] Please report it with the preceding stacktrace at http://github.com/etorreborre/specs2/issues
```
When I attempt to run one of my tests. I'm currently generating fragments, but I've tried converting my test spec to mutable, taking away any import statements, changing how sbt runs the task, and so on.

specs2 version 3.8.4
sbt 0.13
scala version 2.11.8

**Comment by beaushinkle** [comment]
The line that's failing looks to be [this](https://github.com/etorreborre/specs2/blob/2ccbd6a2c69216113c5f392d9b8e4c47b3415ca0/common/src/main/scala/org/specs2/foldm/stream/FoldProcessM.scala#L34) one.
`s.next.run.flatMap(_.getOrElse(Nil).head(t)).run`
`getOrElse(Nil).head` looks like it's having a tough time.

**Comment by etorreborre** [comment]
Ugly indeed. This is now fixed in [3.8.4-20160819060225-dfcb73e](https://oss.sonatype.org/content/repositories/releases/org/specs2/specs2-core_2.11/3.8.4-20160819060225-dfcb73e). It also actually interesting because I am currently working on replacing all that code with something simpler.

**Comment by beaushinkle** [comment]
Thanks!
